### ISSUE TITLE
functional.cond: cannt access frame of function

### ISSUE THREAD

**Issue by FraMecca** [aliak00/ddash#6]
Hi, I have been using your library for a while, especially the cond function that is quite handy.
I have been blocked by this error many times already
I have a struct like with some private members and a function resolve, that contains a nested function parseUrl that uses the cond template:
```
    this(const string content, const string root) @safe
    {
        assert(root.endsWith("/"), root);
        m_content = content;
        m_rooturl = root[0 .. $-1];
    }

    Event[] resolve()
    {
        mixin(scopeInvariant);

        Tuple!(string, string) parseUrl(const string url){
            string src, dst;
            src = url.cond!(
                u => u.startsWith("/"),  {return m_rooturl ~ url;},                     // same error if  u=> m_rooturl ~ u                                                                             
                u => u
    	);

            return tuple!(src, dst);
        }
    // rest of the function

```
I get:
```
/home/user/.dub/packages/ddash-0.0.6/ddash/functional/source/ddash/functional/cond.d(99,25): Error: static function app.ParseEvent.resolve.parseUrl.resolve!(delegate () => this.m_rooturl ~ url).resolve!string.resolve cannot access frame of function app.ParseEvent.resolve.parseUrl
/home/user/.dub/packages/ddash-0.0.6/ddash/functional/source/ddash/functional/cond.d(119,55): Error: template instance `app.ParseEvent.resolve.parseUrl.resolve!(delegate () => this.m_rooturl ~ url).resolve!string` error instantiating
/home/user/git_mio/scarpa/source/app.d(89,22):        instantiated from here: cond!string
/opt/dmd-2.081/import/std/typecons.d(2003,13): Error: static assert:  "Insufficient number of names given."
/home/user/git_mio/scarpa/source/app.d(101,20):        instantiated from here: tuple!()
/usr/bin/dmd failed with exit code 1.
```

This is a more compact test that should avoid that scenario:
```
unittest {
    import std.string;
    struct S{
        string a;
        this(string src){
            a = src;
        }

        void f(){
            auto res = "string".cond!(
                (s) => (s == a), a.startsWith("s"),
                false);                                                                          
    	    assert(res);
        }
    }
    S s = S("string");
    s.f();
}
```

**Comment by FraMecca** [comment]
I want to add that in this cases the solution is a tuple:
```
tuple!("url, "root)(url, root).cond!(
    t => t.url.startsWith("/"), t => t.root ~= t.url,
    t =>t 
);
```

But it greatly pollutes readibility

**Comment by FraMecca** [comment]
To rephrase it,
i want to understand if there is a way to have cond lazily evaluate the second argument, that is the function to be called in case the condition is true.
As it is right now it needs all the arguments to be available at compile time or passed as an argument of the template.

**Comment by aliak00** [comment]
Yo! Glad to hear it's handy :) And thanks for bringing this up. I *think* there may be a way to solve it but I'll have to experiment a bit. Basically, the `cond` template has a `resolver` function that either calls something or just returns it as a literal, and I think D has a problem with accessing frames when you go more than one level deep (so you need two contexts instead of 1). So this is a reduced case:

```
private static template res(alias f) {
    static auto res(T)(T value) {
        return f();
    }
}

template forward(alias f) {
    static auto forward(T)(T value) {
        return res!f(value);
    }
}

unittest {
    auto g() { return 10; }
    forward!g(1);
}
```

I was thinking that inlining the function might work, but then you get this:

```
template forward(alias f) {
    static auto forward(T)(T value) {
        return f();
    }
}

unittest {
    int n = 10;
    auto g() { return 10; }
    forward!(() => n)(1);
}
```

And a similar error :(

I'll dig a bit in to this.

**Comment by aliak00** [comment]
Should be fixed in version 0.0.7. Ping if it's not resolved!

**Comment by FraMecca** [comment]
thank you a lot for fixing this.
